tutor/src/components/add-edit-question/ux.js

```diff
@@ -300,6 +381,103 @@ export default class AddEditQuestionUX {
     this.excludeOriginal = checked;
   }
 
+  @action async publish(shouldExit) {
+    if(!this.hasAnyFeedback) {
+      this.feedbackTipModal = {
+        show: true,
+        shouldExitOnPublish: shouldExit,
+      };
+    }
+    else {
+      this.doPublish(shouldExit);
+    }
+  }
+
+  @action async doPublish(shouldExit) {
+    await this.exercises.createExercise({
+      course: this.course,
+      data: {
+        selectedChapterSection: this.selectedChapterSection.id,
+        authorId: parseInt(this.author.id, 10),
+        derived_from_id: this.from_exercise_id,
+        questionText: this.questionText,
+        questionName: this.questionName,
+        options: this.processOptions(),
+        detailedSolution: this.detailedSolution,
+        isTwoStep: this.isTwoStep,
+        tags: this.processTags(),
+        annonymize: this.annonymize,
+        copyable: this.allowOthersCopyEdit,
+      },
+    });
+
+    if(shouldExit) {
+      this.onDisplayModal(false);
+    }
+    else {
+      this.resetForm();
+    }
+
+    if(this.feedbackTipModal.show) {
+      this.feedbackTipModal = {
+        show: false,
+        shouldExitOnPublish: false,
+      };
+    }
+  }
+
+  processTags() {
+    const tagkeys = ['tagTime', 'tagDifficulty', 'tagBloom', 'tagDok'];
+    let tags = {};
+    forEach(tagkeys, key => {
+      if(this[key]) {
+        tags[key] = { value: this[key].value || this[key] };
+      }
+    });
+
+    return tags;
+  }
+
+  processOptions() {
+    return map(this.filledOptions, o => {
+      return {
+        content: o.text,
+        feedback: o.feedback,
+        correctness: o.isCorrect ? '1.0' : '0.0',
+      };
+    });
+  }
+
+  @action resetForm() {
+    this.questionText = '';
+    this.isTwoStep = false;
+    forEach(this.options, o => {
+      o.text = '';
+      o.feedback = '';
+      o.isCorrect = false;
+    });
+    this.detailedSolution = '';
+    // tags
+    this.tagTime = undefined;
+    this.tagDifficulty = undefined;
+    this.tagBloom = undefined;
+    this.tagDok = undefined;
+    // general
+    this.questionName = '';
+    this.allowOthersCopyEdit = true;
+    this.annonymize = false;
+    this.excludeOriginal = false;
+  }
+
+  @action.bound doExitForm() {
+    if(this.canExit){
+      this.showExitWarningModal = true;
+    }
+    else {
+      this.onDisplayModal(false);
+    }
+  }
+
   /**
    * This method checks if the `fields` (fields in the form) are empty.
    * If so, set this.isEmpty[field] to true.
```

Comment:
I think this'll work fine for now Alex, but typically I find it cleaner to create a model to hold all the data that's being edited, and the UX class would manipulate the model as needed when events occur.

That'd greatly simplify the creation and reset.  Creation could be handled by the model's constructor, and on reset, the UX would throw away that model and create a new one.

Reply:
👍